engine/dlib/src/dlib/opaque_handle_container.h

```diff
@@ -0,0 +1,280 @@
+     *
+     * @name Full
+     * @return [type:bool] Returns true if all object slots are used, and false if there are available slots.
+     */
+    bool Full();
+
+private:
+    T**       m_Objects;
+    uint16_t* m_ObjectVersions;
+    uint32_t  m_Capacity;
+    uint16_t  m_Version;
+
+    // Internal helper functions
+    inline T* GetByIndex(uint32_t i)
+    {
+        assert(i < m_Capacity);
+        return m_Objects[i];
+    }
+
+    uint32_t GetFirstEmptyIndex()
+    {
+        for (int i = 0; i < m_Capacity; ++i)
+        {
+            if (m_Objects[i] == 0)
+            {
+                return i;
+            }
+        }
+        return INVALID_OPAQUE_HANDLE;
+    }
+
+    dmOpaqueHandleContainer(const dmOpaqueHandleContainer<T>&) {}
+    void operator =(const dmOpaqueHandleContainer<T>&) {}
+    void operator ==(const dmOpaqueHandleContainer<T>&) {}
+};
+
+
+//////////////////////////////////////////////////////////////
+// Public API implementation
+//////////////////////////////////////////////////////////////
+
+template <typename T>
+dmOpaqueHandleContainer<T>::dmOpaqueHandleContainer()
+: m_Objects(0)
+, m_Capacity(0)
+, m_Version(0)
+{}
+
+template <typename T>
+dmOpaqueHandleContainer<T>::dmOpaqueHandleContainer(uint32_t inital_capacity)
+: dmOpaqueHandleContainer()
+{
+    Allocate(inital_capacity);
+}
+
+template <typename T>
+dmOpaqueHandleContainer<T>::~dmOpaqueHandleContainer()
+{
+    if (m_Objects != 0)
+    {
+        free(m_Objects);
+        free(m_ObjectVersions);
+    }
+}
+
+template <typename T>
+bool dmOpaqueHandleContainer<T>::Allocate(uint32_t count)
+{
+    if (m_Objects == 0x0)
+    {
+        m_Objects        = (T**) malloc(count * sizeof(sizeof(T*)));
+        m_ObjectVersions = (uint16_t*) malloc(count * sizeof(uint16_t));
+    }
+    else
+    {
+        uint32_t new_capacity = m_Capacity + count;
+        m_Objects             = (T**) realloc(m_Objects, new_capacity * sizeof(T*));
+        m_ObjectVersions      = (uint16_t*) realloc(m_ObjectVersions, new_capacity * sizeof(uint16_t));
+    }
+
+    memset(m_Objects + m_Capacity, 0, count * sizeof(T*));
+    memset(m_ObjectVersions + m_Capacity, 0, count * sizeof(uint16_t));
+
+    m_Capacity += count;
+
+    return m_Objects != 0x0 && m_ObjectVersions != 0x0;
+}
+
+template <typename T>
+T* dmOpaqueHandleContainer<T>::Get(HOpaqueHandle handle)
+{
+    if (handle == 0 || handle == INVALID_OPAQUE_HANDLE)
+    {
+        return 0;
+    }
+
+    uint32_t version = handle >> 16;
+    uint32_t index   = handle & 0xFFFF;
+    T* entry         = GetByIndex(index);
+
+    if (entry == 0 || m_ObjectVersions[index] != version)
+    {
+        return 0;
+    }
+
+    return entry;
+}
+
+template <typename T>
+HOpaqueHandle dmOpaqueHandleContainer<T>::Put(T* obj)
+{
+    uint32_t index = GetFirstEmptyIndex();
+    if (index == INVALID_OPAQUE_HANDLE)
+    {
+        return INVALID_OPAQUE_HANDLE;
+    }
```

Comment:
We don't handle the case when the `m_Capacity==0xFFFF`...

Reply:
Hm, put doesn't grow the container, even if the capacity is the maximum allowed we can still have free slots to put handles.. But maybe the comment wasn't meant regarding this function but more overall?